File: experiments/src/experiment/strategy.jl
module StrategyModule

using ..FeaturesModule
using ..NetworkModule
using ..EvaluateModule
using ..DoActionsModule
using ..OptimizerModule

@kwdef struct Strategy
    base_net::AbstractNetwork
    feats::Vector{AbstractFeature}
    actions::AbstractActions
    penalties::AbstractPenalties
    stop_conds::StopConds
    opt::AbstractOptimizer
    entry_ptr::NodePtr
    exit_ptr::NodePtr
    stop_loss::Float64
    take_profit::Float64
    max_hold_time::Int 
end
export Strategy

@kwdef struct NetworkSignal
    entry::Bool
    exit::Bool
end
export NetworkSignal

function net_signals!(strategy::Strategy, net::AbstractNetwork, feat_matrix::Matrix{Float64}, delay::Int)::Vector{NetworkSignal}

    n_rows = size(feat_matrix, 1)
    
    signals = Vector{NetworkSignal}(undef, n_rows)
    signals[1:delay] .= fill(NetworkSignal(
        entry = false,
        exit = false
    ), delay)

    reset_state!(net)

    for row ∈ delay + 1:n_rows

        eval!(net, feat_matrix, row)
        
        entry_value = node_value(net, strategy.entry_ptr)
        exit_value = node_value(net, strategy.exit_ptr)
        
        signals[row] = NetworkSignal(
            entry = entry_value,
            exit = exit_value
        )
    end

    return signals
end
export net_signals!

end

File: experiments/alphchemy/src/experiment/strategy.rs
use crate::network::network::{Network, NodePtr, Penalties};
use crate::features::features::{Feature, TimestampedTable};
use crate::actions::actions::Actions;
use crate::optimizer::optimizer::StopConds;
use crate::optimizer::genetic::GeneticOpt;
use serde_json::{Value, json};
#[cfg(test)]
use mockall::automock;

#[derive(Clone, Debug)]
pub struct NetSignals {
    pub entry: bool,
    pub exit: bool
}

pub struct Strategy<T: Network, P: Penalties<T>, A: Actions<T>> {
    pub base_net: T,
    pub feats: Vec<Feature>,
    pub actions: A,
    pub penalties: P,
    pub stop_conds: StopConds,
    pub opt: GeneticOpt,
    pub entry_ptr: NodePtr,
    pub exit_ptr: NodePtr,
    pub stop_loss: f64,
    pub take_profit: f64,
    pub max_hold_time: usize,
    pub qty: f64
}

#[cfg_attr(test, automock)]
trait StrategyDeps<T: Network> {
    fn reset_state(&self, net: &mut T) {
        net.reset_state();
    }

    fn eval(&self, net: &mut T, feat_table: &TimestampedTable, row: usize) {
        net.eval(feat_table, row);
    }

    fn node_value(&self, net: &T, node_ptr: &NodePtr) -> bool {
        net.node_value(node_ptr)
    }
}

struct StrategyDepsImpl;
impl<T: Network> StrategyDeps<T> for StrategyDepsImpl {}

impl<T: Network, P: Penalties<T>, A: Actions<T>> Strategy<T, P, A> {
    pub fn to_json(&self) -> Value {
        json!({
            "base_net": self.base_net.to_json(),
            "feats": self.feats,
            "actions": self.actions.to_json(),
            "penalties": self.penalties.to_json(),
            "stop_conds": self.stop_conds,
            "opt": self.opt.to_json(),
            "entry_ptr": self.entry_ptr,
            "exit_ptr": self.exit_ptr,
            "stop_loss": self.stop_loss,
            "take_profit": self.take_profit,
            "max_hold_time": self.max_hold_time,
            "qty": self.qty
        })
    }

    fn _net_signals<D>(&self, deps: &D, net: &mut T, feat_table: &TimestampedTable, start_idx: usize, end_idx: usize, delay: usize) -> Vec<NetSignals> where D: StrategyDeps<T> {
        let n_rows = end_idx - start_idx + 1;
        let mut signals = Vec::with_capacity(n_rows);

        for _ in 0..delay {
            signals.push( NetSignals {
                entry: false,
                exit: false
            });
        }

        deps.reset_state(net);

        for i in delay..n_rows {
            let row_idx = start_idx + i - delay;
            deps.eval(net, feat_table, row_idx);

            let entry = deps.node_value(net, &self.entry_ptr);
            let exit = deps.node_value(net, &self.exit_ptr);
            let new_signals = NetSignals {
                entry,
                exit
            };
            signals.push(new_signals);
        }

        signals
    }

    pub fn net_signals(&self, net: &mut T, feat_table: &TimestampedTable, start_idx: usize, end_idx: usize, delay: usize) -> Vec<NetSignals> {
        self._net_signals(&StrategyDepsImpl, net, feat_table, start_idx, end_idx, delay)
    }
}
